xdsl/dialects/csl.py

```diff
@@ -344,20 +404,121 @@ def parse(cls, parser: Parser) -> FuncOp:
         return func
 
     def print(self, printer: Printer):
-        print_func_op_like(
-            printer,
-            self.sym_name,
-            self.function_type,
-            self.body,
-            self.attributes,
-            arg_attrs=self.arg_attrs,
-            reserved_attr_names=(
-                "sym_name",
-                "function_type",
-                "sym_visibility",
-                "arg_attrs",
-            ),
+        FuncBase._print(self, printer)
+
+
+@irdl_op_definition
+class TaskOp(IRDLOperation, FuncBase):
+    """
+    Represents a task in CSL. All three types of task are represented by this Op.
+
+    It carries the ID it should be bound to, in case of local and control tasks
+    this is the task ID, in the case of the data task, it's the id of the color
+    the task is bound to.
+
+    NOTE: Control tasks not yet implemented
+    """
+
+    name = "csl.task"
+
+    kind = prop_def(TaskKindAttr)
+    id = prop_def(IntegerAttr[IntegerType])
+
+    traits = frozenset([InModuleKind(ModuleKind.PROGRAM)])
+
+    def __init__(
+        self,
+        name: str,
+        function_type: FunctionType | tuple[Sequence[Attribute], Attribute | None],
+        region: Region | type[Region.DEFAULT] = Region.DEFAULT,
+        *,
+        task_kind: TaskKindAttr | TaskKind,
+        arg_attrs: ArrayAttr[DictionaryAttr] | None = None,
+        res_attrs: ArrayAttr[DictionaryAttr] | None = None,
+        id: IntegerAttr[IntegerType] | int,
```

Comment:
I think task kind and task ID should be optional, right? Can tasks not have IDs?

Reply:
Kind enforces restrictions on what args the task can have, I think it should be mandatory.

The id I was using to immediately bind tasks upon declaration, but I guess this can be done at runtime too.

I can make it optional and only bind the task if id is present (in the printer).

This would necessitate some sort of `csl.bind_task` op, which we don't currently have, but it'd be pretty simple to add.

Reply:
Hmm, whatever you prefer. Just wanted to note this. Is there overlap in what args are allowed?

Reply:
Local: no args
Data: 1 - 4 args of no more than 64 bits total
Control: 0 - 4 args of no more than 64 bits total

The way the meaning of the args also changes between data and control, control tasks get passed their IDs, data tasks just get the data.